src/components/Playground.svelte

```diff
@@ -0,0 +1,201 @@
+      }, wait);
+    };
+  };
+
+  type State = "initializing" | "compiling" | "compiledSuccessfully" | "compiledWithErrors";
+
+  function getLoadingState(state: State): string | null {
+    switch (state) {
+      case "initializing":
+        return "Initializing";
+      case "compiling":
+        return "Compiling";
+      default:
+        return null;
+    }
+  }
+
+  let output = "";
+  let state: State = "initializing";
+  $: loadingState = getLoadingState(state);
+
+  let themeChangeUnsub: () => void;
+  let worker: Worker;
+  let editor: monaco.editor.IStandaloneCodeEditor;
+
+  function compile() {
+    state = "compiling";
+
+    worker.postMessage({
+      content: editor.getValue(),
+    });
+
+    worker.onmessage = ({ data }) => {
+      if (data?.stderr) {
+        state = "compiledWithErrors";
+        output = data.stderr;
+      } else {
+        state = "compiledSuccessfully"
+        if (data?.stdout) {
+          output = data.stdout;
+        } else {
+          output = "< no program output >";
+        }
+      }
+    };
+
+    worker.onerror = (err) => {
+      console.error(err);
+    };
+
+    editor.layout();
+  }
+
+  let examplesToggled = false;
+  function toggleExamples(): void {
+    examplesToggled = !examplesToggled;
+  }
+
+  let showCopiedMessage = false;
+  let existingCopyMessageTimeout: NodeJS.Timeout;
+  function copyCode(): void {
+    navigator.clipboard.writeText(window.location.href);
+    showCopiedMessage = true;
+    clearTimeout(existingCopyMessageTimeout);
+    existingCopyMessageTimeout = setTimeout(() => showCopiedMessage = false, 1000);
+  }
+
+  const switchCodeExample = (compressedCode: string) => () => {
+    window.history.pushState(null, "", "?code=" + compressedCode);
+    editor.setValue(lzString.decompressFromEncodedURIComponent(compressedCode));
+  }
+
+  onMount(async () => {
+    worker = new CompilerWorker();
+    worker.onerror = (err) => console.error(err);
+
+    const urlParams = new URLSearchParams(window.location.search);
+    const code = lzString.decompressFromEncodedURIComponent(urlParams.get("code")!)
+      || 'module Playground\n\nprint("Hello, world!")\n';
+    editor = await createGrainEditor("editor", code);
+
+    const prefersDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
+    themeChangeUnsub = store.$selectedTheme.subscribe(selectedTheme => {
+      monaco.editor.setTheme(selectedTheme === "light" || (!selectedTheme && !prefersDark) ? "github-light" : "github-dark-modified");
+    });
+
+    worker.addEventListener("message", ({ data }) => {
+      if (data?.initialized) {
+        compile();
+      }
+    });
+
+    editor.onDidChangeModelContent(() => {
+      const code = editor.getValue();
+      const compressed = lzString.compressToEncodedURIComponent(code);
+      const searchParams = new URLSearchParams({
+        code: compressed
+      });
+      window.history.pushState(null, "", "?" + searchParams.toString());
```

Comment:
I think it would make sense to replace this with `replaceState`, using `pushState` makes it very hard to go back from the playground.